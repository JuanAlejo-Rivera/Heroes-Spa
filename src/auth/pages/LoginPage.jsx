import React, { useContext } from 'react'
import { useNavigate } from 'react-router-dom'
import { AuthContext } from '../context/AuthContext'

export const LoginPage = () => {

  const {login} = useContext( AuthContext )
  
  const navigate = useNavigate()
  const onLogin = () => {

    // Cuando le damos click al loguin navegaremos al valor guardado en el local stotage
    // Si no hay valor guardado refresamos al / que seria marvel
    const lastPath = localStorage.getItem('lastPath') || '/';

    login('Alejandro Rivera')

    navigate(lastPath, {
      replace: true, //indica que la nueva ruta (lastPath) reemplazar√° a la actual en el historial del navegador
    })
  }


  return (
    <div className='container'>

      <div className='contarine mt-5'>
        <h1>Login</h1>
        <hr />
        <button
          className='btn btn-primary'
          onClick={onLogin}
        >
          Login
        </button>
      </div>
    </div>
  )
}
