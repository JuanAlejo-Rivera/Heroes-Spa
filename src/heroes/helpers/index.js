export * from './getHeroesByPublisher'
export * from './getHeroByid'
export * from './getHerosByName'