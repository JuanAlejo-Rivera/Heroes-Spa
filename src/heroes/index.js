export * from './pages';
export * from './routes/HeroesRutas'