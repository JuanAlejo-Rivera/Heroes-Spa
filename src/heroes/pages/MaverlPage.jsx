import React from "react"
import { HeroeList } from "../components"


export const MaverlPage = () => {
  return (
    <>
      <h1>MaverlPage</h1>
      <hr />

      <HeroeList publisher={'Marvel Comics'}/>

    </>
  )
}
