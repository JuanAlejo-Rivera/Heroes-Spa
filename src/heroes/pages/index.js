export * from './DCPage';
export * from './MaverlPage';
export * from './SearchPage'
export * from './HeroPages'