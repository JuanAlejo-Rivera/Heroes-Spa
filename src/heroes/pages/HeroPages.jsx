import { Navigate, useNavigate, useParams } from "react-router-dom";
import { getHeroByid } from "../helpers";
import React, { useMemo } from "react";


export const HeroPages = () => {

  const navigate = useNavigate()
  const { id } = useParams();
  
  //Use memo para memorizar valores y useCallback para memorizar funciones
  const hero = useMemo( () => getHeroByid(id), [ id ]) 

  const onNavigateback = () => {

    // if (hero.publisher.match('DC Comics')) {
    //   navigate('/dc', {
    //     replace: true
    //   })
    // }
    // if (hero.publisher.match('Marvel Comics')) {
    //   navigate('/marvel', {
    //     replace: true
    //   })
    // }


    navigate(-1); //Regresa al a pagina anterior, puede sacarnos de la app

  }


  if (!hero) {
    return <Navigate to="/marvel" />
  }



  return (
    <div className="row mt-5">
      <div className="col-4">
        <img
          src={`/heroes/${id}.jpg`}
          alt={hero.superhero}
          className="img-thumbnail animate__animated animate__fadeInLeft"
        />
      </div>


      <div className="col-8">
        <h3>{hero.superhero}</h3>
        <ul className="list-group list-group-flush">
          <li className="list-group-item"> <b>Alter ego:</b> {hero.alter_ego} </li>
          <li className="list-group-item"> <b>Publisher:</b> {hero.publisher} </li>
          <li className="list-group-item"> <b>First appearance:</b> {hero.first_appearance} </li>
        </ul>

        <h5 className="mt-3"> Characters</h5>
        <p>{hero.characters}</p>

        <button
          aria-label="buttonNavigate"
          onClick={onNavigateback}
          className="btn btn-dark"
        >
          Back
        </button>
      </div>
    </div>
  )
}
