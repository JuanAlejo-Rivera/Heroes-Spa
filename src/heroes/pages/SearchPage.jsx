import { useLocation, useNavigate } from "react-router-dom";
import queryString from "query-string";
import { useForm } from "../../hook/useForm";
import { HeroCard } from "../components";
import { getHerosByName } from "../helpers";

export const SearchPage = () => {

  const navigate = useNavigate();
  const location = useLocation()
  // console.log(location)
  const { q = '' } = queryString.parse(location.search); // se instalo la libreria npm install query-string
  // console.log(query)
  //Los queryParameter son opcionales

  const heroes = getHerosByName(q);

  const showSearch = (q.length === 0) //Si la presona no escribe nada es T de lo contrario F
  const showError  = (q.length !== 0 && heroes.length === 0)


  const { formState, onInputChange, onResetForm, SearchText } = useForm({
    SearchText: ''
    // SearchText: q //Si quiero que lo escrito en el input no se borre
  });

  const onSearchSubmit = (event) => {
    event.preventDefault();
    // if (SearchText.trim().length <= 1) return;

    navigate(`?q=${SearchText}`)
  }

  return (
    <>
      <h1>Search</h1>
      <hr />

      <div className="row">

        <div className="col-5">
          <h4>Searching</h4>
          <hr />
          <form onSubmit={onSearchSubmit}>
            <input type="text"
              placeholder="Search a hero"
              className="form-control"
              name="SearchText"
              autoComplete="off"
              value={SearchText} //necesario para conectar
              onChange={onInputChange} //necesario para conectar
            />

            <button className="btn btn-outline-primary mt-1">
              Search
            </button>
          </form>
        </div>



        <div className="col-7">
          <h4>Results</h4>
          <hr />
          {/* {
            (q === '' ?
              <div className="alert alert-primary">Search a hero</div>
              :(heroes.length === 0) &&
              <div className="alert alert-danger">No hero with <b>{q}</b></div>
            )
          } */}

          <div className="alert alert-primary animate__animated animate__fadeIn" style={{display: showSearch ? " ": 'none'}}>
            Search a hero
          </div>

          <div className="alert alert-danger animate__animated animate__fadeIn" style={{display: showError ? " ": 'none' }}>
            No hero with <b>{q}</b>
            </div>

          {
            heroes.map(hero => (
              <HeroCard key={hero.id} {...hero} />
            ))
          }

        </div>
      </div>


    </>
  )
}



