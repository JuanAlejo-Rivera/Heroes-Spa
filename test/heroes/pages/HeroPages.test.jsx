

import { fireEvent, render, screen } from "@testing-library/react";
import React from "react";
import { MemoryRouter, Route, Routes } from "react-router-dom";
import { HeroPages } from "../../../src/heroes/pages/HeroPages";
import { getHeroByid } from "../../../src/heroes/helpers";

// Mock de la función useNavigate
const mockedUseNavigate = jest.fn();

// Simula las funciones del router
jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockedUseNavigate
}));

// Simula la función getHeroByid
jest.mock('../../../src/heroes/helpers', () => ({
    getHeroByid: jest.fn(),
}));

describe('Pruebas en el HeroPages', () => {

    /*
Nota: si no simulamos el componente trae los datos ReadableStream
pero eso seria depender de componentes externos como API's, datos importados o de otros archivos
es mejor realizar la simunalcion de la respuesta de datos como acontinuacion
*/

    // Limpia los mocks antes de cada prueba y simula los datos de getHeroByid
    beforeEach(() => {
        jest.clearAllMocks();
        getHeroByid.mockReturnValue({
            'superhero': 'Spider Man-prueba',
            'publisher': 'Marvel Comics-prueba',
            'alter_ego': 'Peter Parker-prueba',
            'first_appearance': 'Amazing Fantasy #15-prueba',
            'characters': 'Peter Parker-prueba'
        });
    });

    test('Se debe comprobar el boton de back', () => {
        // Renderiza el componente con la URL inicial y las rutas necesarias
        const {container} = render(
            <MemoryRouter initialEntries={['/hero/marvel-spider']}>
                <Routes>
                    <Route path="/hero/:id" element={<HeroPages />} />
                    <Route path="/marvel" element={'Estoy en marvel'} />
                </Routes>
            </MemoryRouter>
        );
        expect(container).toMatchSnapshot();
        // screen.debug();
        // Encuentra el botón y simula un clic
        const button = screen.getByRole('button');
        fireEvent.click(button);

        // Pruebas
        expect(mockedUseNavigate).toHaveBeenCalledWith(-1);
        expect(screen.getAllByText(/Peter Parker-prueba/i)).toBeTruthy()
        expect(screen.getAllByText(/Peter Parker-prueba/i)).toHaveLength(2);

    });


    test('Si no se retorna un heroe valido debe retornar a marvel', () => { 
        
        render(
            <MemoryRouter initialEntries={['/marvel']}>
                <Routes>
                    <Route path="/hero/:id" element={<HeroPages />} />
                    <Route path="/marvel" element={'Estoy en marvel'} />
                </Routes>
            </MemoryRouter>
        );
        screen.debug()
        // expect(screen.getAllByText(/Estoy en marvel/i)).toBeTruthy()
     });
});
