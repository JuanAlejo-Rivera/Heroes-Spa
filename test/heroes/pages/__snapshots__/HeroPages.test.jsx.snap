// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el HeroPages Se debe comprobar el boton de back 1`] = `
<div>
  <div
    class="row mt-5"
  >
    <div
      class="col-4"
    >
      <img
        alt="Spider Man-prueba"
        class="img-thumbnail animate__animated animate__fadeInLeft"
        src="/heroes/marvel-spider.jpg"
      />
    </div>
    <div
      class="col-8"
    >
      <h3>
        Spider Man-prueba
      </h3>
      <ul
        class="list-group list-group-flush"
      >
        <li
          class="list-group-item"
        >
           
          <b>
            Alter ego:
          </b>
           
          Peter Parker-prueba
           
        </li>
        <li
          class="list-group-item"
        >
           
          <b>
            Publisher:
          </b>
           
          Marvel Comics-prueba
           
        </li>
        <li
          class="list-group-item"
        >
           
          <b>
            First appearance:
          </b>
           
          Amazing Fantasy #15-prueba
           
        </li>
      </ul>
      <h5
        class="mt-3"
      >
         Characters
      </h5>
      <p>
        Peter Parker-prueba
      </p>
      <button
        aria-label="buttonNavigate"
        class="btn btn-dark"
      >
        Back
      </button>
    </div>
  </div>
</div>
`;
